php/libraries/NDB_BVL_Instrument.class.inc

```diff
@@ -1523,6 +1523,40 @@ abstract class NDB_BVL_Instrument extends NDB_Page
         return null;
     }
 
+    /**
+     * Determines what percentage of the data entry has been completed, based on the
+     * number of fields in $_requiredElements having data
+     *
+     * @return int the data entry completion percentage
+     */
+    function determineDataEntryCompletionProgress(): int
+    {
+        // don't bother checking anything if the required elements array is empty
+        if (empty($this->_requiredElements)) {
+            return 100;
+        }
+        $allData       = $this->loadInstanceData($this);
+        $unanswered    = 0;
+        $totalElements = count($this->_requiredElements);
+        foreach ($this->_requiredElements as $field) {
+            // consider status field if not_answered chosen as an answer
+            $statusField = $field . '_status';
+            // field shouldn't be just empty() b/c 0 is a valid
+            // value for some required fields
+            if ((is_null($allData[$field]) || $allData[$field] === "")
```

Comment:
```suggestion
            if ((($allData[$field] ?? null) === null) || (($allData[$field] ?? "") === "")
```

When using this in the other PR, it generated index undefined errors for some fields which resulted in invalid JSON being returned. This fixes it.